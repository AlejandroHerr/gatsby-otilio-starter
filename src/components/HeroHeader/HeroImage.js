// @flow
/* globals document */
import React, { PureComponent } from 'react';
import Img from 'gatsby-image';

import type { SizeImageSharpType } from '../../types/gatsby';

import styles from './HeroHeader.module.scss';

type PropsType = {
  image: SizeImageSharpType,
};

type StateType = {
  translateY: number,
};

export default class HeroImage extends PureComponent<PropsType, StateType> {
  handleScroll: () => void;

  constructor(props: PropsType) {
    super(props);

    this.state = {
      translateY: 0,
    };

    this.handleScroll = this.handleScroll.bind(this);
  }

  componentDidMount() {
    document.addEventListener('scroll', this.handleScroll);
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.handleScroll);
  }

  handleScroll() {
    const scrollTop = (document.documentElement && document.documentElement.scrollTop)
      || 0;

    this.setState({ translateY: scrollTop / 5 });
  }
  render() {
    const { image } = this.props;
    const { translateY } = this.state;

    return (
      <div
        className={styles.hero_image}
        style={{ transform: `translate3d(0px, ${translateY}px, 0px)` }}
      >
        <Img
          sizes={image}
          style={{
            position: 'absolute',
            left: 0,
            top: 0,
            width: '100%',
            height: '100%',
            margin: '0',
          }}
        />
        <div className={styles.hero_image__shadow} />
      </div>
    );
  }
}
