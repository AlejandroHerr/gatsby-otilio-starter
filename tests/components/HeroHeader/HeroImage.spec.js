/* globals document */
import React from 'react';
import { mount } from 'enzyme';

import HeroImage from '../../../src/components/HeroHeader/HeroImage';

const defaultProps = {
  image: {
    src: 'src',
  },
};

const setup = (props = {}) => {
  const heroImage = mount(<HeroImage {...defaultProps} {...props} />);

  return {
    heroImage,
  };
};


describe('HeroCover', () => {
  describe('HeroImage', () => {
    it('should render a HeroImage', () => {
      const { heroImage } = setup();

      expect(heroImage.find('div').first().prop('style').transform).toBe('translate3d(0px, 0px, 0px)');

      expect(heroImage).toMatchSnapshot();
    });
    it('should listen to scroll events', () => {
      document.addEventListener = jest.fn();
      document.removeEventListener = jest.fn();

      const { heroImage } = setup();
      const { handleScroll } = heroImage.instance();

      expect(document.addEventListener).toHaveBeenCalledWith('scroll', handleScroll);

      heroImage.unmount();

      expect(document.removeEventListener).toHaveBeenCalledWith('scroll', handleScroll);
    });
    it('should handle on scroll events', () => {
      const { heroImage } = setup();

      const scrollTop = 50;

      document.documentElement.scrollTop = scrollTop;

      heroImage.instance().handleScroll();
      heroImage.update();

      expect(heroImage.find('div').first().prop('style').transform).toBe(`translate3d(0px, ${scrollTop / 5}px, 0px)`);
      expect(heroImage).toMatchSnapshot();

      document.documentElement.scrollTop = null;

      heroImage.instance().handleScroll();
      heroImage.update();

      expect(heroImage.find('div').first().prop('style').transform).toBe('translate3d(0px, 0px, 0px)');
      expect(heroImage).toMatchSnapshot();
    });
  });
});
