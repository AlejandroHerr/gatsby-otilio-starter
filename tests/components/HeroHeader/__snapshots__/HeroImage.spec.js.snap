// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroCover HeroImage should handle on scroll events 1`] = `
<HeroImage
  image={
    Object {
      "src": "src",
    }
  }
>
  <div
    className="hero_image"
    style={
      Object {
        "transform": "translate3d(0px, 10px, 0px)",
      }
    }
  >
    <Image
      Tag="div"
      alt=""
      fadeIn={true}
      sizes={
        Object {
          "src": "src",
        }
      }
      style={
        Object {
          "height": "100%",
          "left": 0,
          "margin": "0",
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    >
      <div
        className=" gatsby-image-outer-wrapper"
        style={
          Object {
            "position": "initial",
          }
        }
      >
        <div
          className=" gatsby-image-wrapper"
          style={
            Object {
              "height": "100%",
              "left": 0,
              "margin": "0",
              "overflow": "hidden",
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "paddingBottom": "NaN%",
                "width": "100%",
              }
            }
          />
          <Img
            alt=""
            onLoad={[Function]}
            src="src"
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <img
              alt=""
              onLoad={[Function]}
              src="src"
              style={
                Object {
                  "height": "100%",
                  "left": 0,
                  "objectFit": "cover",
                  "objectPosition": "center",
                  "opacity": 1,
                  "position": "absolute",
                  "top": 0,
                  "transition": "opacity 0.5s",
                  "width": "100%",
                }
              }
            />
          </Img>
          <noscript
            dangerouslySetInnerHTML={
              Object {
                "__html": "<img width=\\"\\" height=\\"\\" src=\\"src\\" srcset=\\"undefined\\" alt=\\"\\" title=\\"\\" sizes=\\"\\" style=\\"position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/>",
              }
            }
          />
        </div>
      </div>
    </Image>
    <div
      className="hero_image__shadow"
    />
  </div>
</HeroImage>
`;

exports[`HeroCover HeroImage should handle on scroll events 2`] = `
<HeroImage
  image={
    Object {
      "src": "src",
    }
  }
>
  <div
    className="hero_image"
    style={
      Object {
        "transform": "translate3d(0px, 0px, 0px)",
      }
    }
  >
    <Image
      Tag="div"
      alt=""
      fadeIn={true}
      sizes={
        Object {
          "src": "src",
        }
      }
      style={
        Object {
          "height": "100%",
          "left": 0,
          "margin": "0",
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    >
      <div
        className=" gatsby-image-outer-wrapper"
        style={
          Object {
            "position": "initial",
          }
        }
      >
        <div
          className=" gatsby-image-wrapper"
          style={
            Object {
              "height": "100%",
              "left": 0,
              "margin": "0",
              "overflow": "hidden",
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "paddingBottom": "NaN%",
                "width": "100%",
              }
            }
          />
          <Img
            alt=""
            onLoad={[Function]}
            src="src"
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <img
              alt=""
              onLoad={[Function]}
              src="src"
              style={
                Object {
                  "height": "100%",
                  "left": 0,
                  "objectFit": "cover",
                  "objectPosition": "center",
                  "opacity": 1,
                  "position": "absolute",
                  "top": 0,
                  "transition": "opacity 0.5s",
                  "width": "100%",
                }
              }
            />
          </Img>
          <noscript
            dangerouslySetInnerHTML={
              Object {
                "__html": "<img width=\\"\\" height=\\"\\" src=\\"src\\" srcset=\\"undefined\\" alt=\\"\\" title=\\"\\" sizes=\\"\\" style=\\"position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/>",
              }
            }
          />
        </div>
      </div>
    </Image>
    <div
      className="hero_image__shadow"
    />
  </div>
</HeroImage>
`;

exports[`HeroCover HeroImage should render a HeroImage 1`] = `
<HeroImage
  image={
    Object {
      "src": "src",
    }
  }
>
  <div
    className="hero_image"
    style={
      Object {
        "transform": "translate3d(0px, 0px, 0px)",
      }
    }
  >
    <Image
      Tag="div"
      alt=""
      fadeIn={true}
      sizes={
        Object {
          "src": "src",
        }
      }
      style={
        Object {
          "height": "100%",
          "left": 0,
          "margin": "0",
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    >
      <div
        className=" gatsby-image-outer-wrapper"
        style={
          Object {
            "position": "initial",
          }
        }
      >
        <div
          className=" gatsby-image-wrapper"
          style={
            Object {
              "height": "100%",
              "left": 0,
              "margin": "0",
              "overflow": "hidden",
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "paddingBottom": "NaN%",
                "width": "100%",
              }
            }
          />
          <Img
            alt=""
            onLoad={[Function]}
            src="src"
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <img
              alt=""
              onLoad={[Function]}
              src="src"
              style={
                Object {
                  "height": "100%",
                  "left": 0,
                  "objectFit": "cover",
                  "objectPosition": "center",
                  "opacity": 1,
                  "position": "absolute",
                  "top": 0,
                  "transition": "opacity 0.5s",
                  "width": "100%",
                }
              }
            />
          </Img>
          <noscript
            dangerouslySetInnerHTML={
              Object {
                "__html": "<img width=\\"\\" height=\\"\\" src=\\"src\\" srcset=\\"undefined\\" alt=\\"\\" title=\\"\\" sizes=\\"\\" style=\\"position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/>",
              }
            }
          />
        </div>
      </div>
    </Image>
    <div
      className="hero_image__shadow"
    />
  </div>
</HeroImage>
`;
